pkg/helm/helm3lib/helm3lib.go

```diff
@@ -0,0 +1,315 @@
+	"github.com/flant/addon-operator/pkg/utils"
+
+	klient "github.com/flant/kube-client/client"
+)
+
+// Init runs
+func Init(opts *Options) error {
+	hc := &LibClient{
+		LogEntry: log.WithField("operator.component", "helm3lib"),
+	}
+	options = opts
+	err := hc.initAndVersion()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Library use client
+type LibClient struct {
+	KubeClient klient.Client
+	LogEntry   *log.Entry
+	Namespace  string
+}
+
+type Options struct {
+	Namespace  string
+	HistoryMax int32
+	Timeout    time.Duration
+	KubeClient klient.Client
+}
+
+var _ client.HelmClient = &LibClient{}
+var options *Options
+var actionConfig *action.Configuration
+
+func NewClient(logLabels ...map[string]string) client.HelmClient {
+	logEntry := log.WithField("operator.component", "helm3lib")
+	if len(logLabels) > 0 {
+		logEntry = logEntry.WithFields(utils.LabelsToLogFields(logLabels[0]))
+	}
+
+	return &LibClient{
+		LogEntry:   logEntry,
+		KubeClient: options.KubeClient,
+		Namespace:  options.Namespace,
+	}
+}
+
+func (h *LibClient) CommandEnv() []string {
+	res := make([]string, 0)
+	return res
+}
+
+func (h *LibClient) WithKubeClient(client klient.Client) {
+	h.KubeClient = client
+}
+
+// initAndVersion runs helm version command.
+func (h *LibClient) initAndVersion() error {
+	ac := new(action.Configuration)
+
+	env := cli.New()
+
+	err := ac.Init(env.RESTClientGetter(), options.Namespace, "secrets", h.LogEntry.Debugf)
+	if err != nil {
+		return err
+	}
+
+	actionConfig = ac
+
+	log.Infof("Helm 3 version: %s", chartutil.DefaultCapabilities.HelmVersion.Version)
+
+	return nil
+}
+
+func (h *LibClient) DeleteSingleFailedRevision(releaseName string) error {
+	// No need to delete single failed revision anymore
+	// https://github.com/helm/helm/issues/8037#issuecomment-622217632
+	return nil
+}
+
+func (h *LibClient) DeleteOldFailedRevisions(releaseName string) error {
+	// No need to delete single failed revision anymore
+	// https://github.com/helm/helm/issues/8037#issuecomment-622217632
+	return nil
+}
+
+// LastReleaseStatus returns last known revision for release and its status
+func (h *LibClient) LastReleaseStatus(releaseName string) (revision string, status string, err error) {
+	release, err := actionConfig.Releases.Last(releaseName)
+	if err != nil {
+		if err == driver.ErrReleaseNotFound || strings.HasPrefix(err.Error(), "no revision for release") {
+			return "0", "", fmt.Errorf("release '%s' not found\n", releaseName)
+		}
+		return "", "", err
+	}
+
+	return strconv.FormatInt(int64(release.Version), 10), release.Info.Status.String(), nil
+}
+
+func (h *LibClient) UpgradeRelease(releaseName string, chartName string, valuesPaths []string, setValues []string, namespace string) error {
+	upg := action.NewUpgrade(actionConfig)
+	if namespace != "" {
+		upg.Namespace = namespace
+	}
+
+	upg.Install = true
+	upg.MaxHistory = int(options.HistoryMax)
+	upg.Timeout = options.Timeout
```

Comment:
Just a thought... Shouldn't we enable Wait for helm3lib by default?

Reply:
Don't think so. Cause that way any broken deployment/daemonset/etc in CrashLoopBackoff status in any module could block the queue. And it will make deployment much more longer

Reply:
Oops. I thought wait is fail-fast.